### DISCUSSION TITLE
Plugin execution order: service vs. route

### DISCUSSION THREAD

**Discussion by shawnchapla-8451** [Kong/kong#10999] [Help]
We have a couple of custom plugins which need to execute in a particular order.  The one that needs to execute first is configured on a specific route while the other is on the service (to cover the rest of the routes).  Both only run in the access phase, and following the documentation [here](https://docs.konghq.com/gateway/latest/plugin-development/custom-logic/#plugins-execution-order), the plugin on the route has a higher number for its static priority than that on the service plugin.

However, from viewing the debug messages in the log, the service plugin is still executing first (because the route plugin would effectively terminate the request, if it were running first, then in theory, the service plugin should never execute.  Or is that an incorrect assumption?

By "terminate the request", the route plugin (written in Python) is performing an exit, but it's never being called because the service plugin is the only one logging anything and it's performing the request termination.

**Comment by shawnchapla-8451** [comment]
I should add that this is Kong OSS (so no dynamic plugin ordering) v3.1.1.

**Comment by shawnchapla-8451** [comment]
Some additional info ... neither route is using any regex:

1. /simpleapi
2. /simpleapi/docs (the plugins on this route should take precedence over that on the service)

As a sanity check, I moved the service plugin to be on route #1, but it's still the one being invoked when I call route #2.  So it's apparently not a service vs. route question.  So to see what would happen, I disabled the plugin now on route #1, but a call to route #2 still doesn't trigger its plugins.  So I'm going to assume I'm doing something wrong in my configuration so will close this discussion.

**Comment by shawnchapla-8451** [comment]
So I did find that the route matching needed to be configured differently for it to match, so that's resolved.  But I still wanted to sanity check something ...

If you have a plugin defined on the service and a different plugin on a route, it appears Kong uses the plugin priority values and, if that of the service's is higher, it runs that before any lower priority route plugins.  Is that correct?

**Comment by shettyh** [comment]
>If you have a plugin defined on the service and a different plugin on a route, it appears Kong uses the plugin priority values and, if that of the service's is higher, it runs that before any lower priority route plugins. Is that correct?

Yes thats correct
